File: day3/day3.py
def loaddata(filename):
    data = []
    with open(filename, "r") as f:
        for line in f.readlines():
            data.append(line.strip())
        f.close()
    return data


def findtrees(data, startx, starty, diffx, diffy):
    pos_x = startx
    pos_y = starty
    num = 0
    rows = len(data)
    columns = len(data[0])
    while pos_y < rows:
        cur_row = data[pos_y]
        if cur_row[pos_x % columns] == '#':
            num = num + 1
        pos_x = pos_x + diffx
        pos_y = pos_y + diffy
    return num


def main():
    data = loaddata("input.txt")
    print(len(data), " rows", len(data[0]), " columns")
    # print(data)
    answer = 1
    slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
    for (diffx, diffy) in slopes:
        num = findtrees(data, 0, 0, diffx, diffy)
        print(num, " trees for %d, %d" % (diffx, diffy))
        answer = answer * num
    print("The answer is ", answer)


if __name__ == "__main__":
    main()

File: day3/day3.jl
function readdata(filename)
    data = []
  
    width = 0
    height = 0
    open(filename, "r") do f
        for line in eachline(f)
            for idx in eachindex(line)
                ch = line[idx]
                if (height == 0) 
                    width = width + 1
                end
                if (ch == '.')
                    append!(data, 0)
                elseif (ch == '#')
                    append!(data, 1)
                end
            end
            height = height + 1
        end
    end
#    println(data) 
    println(width, "x", height)
    transpose(reshape(data, (width, height)))
end
function calctrees(data, sx, sy, dx, dy)
    (rows, cols) = size(data)
    ox = sx
    oy = sy
    count = 0
    while oy <= rows
        count = count + data[oy, 1 + (ox - 1) % cols]
        ox = ox + dx
        oy = oy + dy
    end
    count
end

function main()
    data = readdata("input.txt")
    (rows, cols) = size(data)
    println("Rows $rows Columns $cols")
    slopes = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
    answer = 1
    for (dx, dy) in slopes
        count = calctrees(data, 1, 1, dx, dy)
        println("$count trees for $dx, $dy")
        answer = answer * count
    end
    println("Cumulative Product = $answer")
end

main()